### PR TITLE
Big picture

### PR DESCRIPTION
**Pull request by wycats**
This PR updates the spec with the latest thinking based on discussions with @jonathandturner

### CODE REVIEW

File: interop/reusability.md [hunk 1]
```diff
@@ -0,0 +1,1088 @@
...
+Next, we'll explore how this definition works on each kind of
+declarative property creation in JavaScript.
+
+## Basic Rules of Decorators
+
+Decorators always operate on a particular syntactic element,
+providing a hook into the runtime semantics for that syntax.
+
+If the runtime semantics for the syntax include "let x be the
+result of evaluating *SomeExpression*`, that expression is passed
```
Comment (jayphelps) [inline interop/reusability.md:55]: I believe this is mismatched tick/double quote. Probably was suppose to be either "let x be the result of evaluating _SomeExpression_" or `let x be the result of evaluating *SomeExpression*`

File: interop/private-state.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
++ // true for private methods (#name() {}), false for public methods (name() {})
++ slot: boolean
+}
+
+interface AccessorDecoratorDescriptor {
+  type: string,             // 'accessor'
+  hint: string,             // 'getter', 'setter', 'both'
+  enumerable: boolean,      // default: true
+  wrtiable: boolean,        // default: true
```
Comment (jayphelps) [inline interop/private-state.md:89]: s/wrtiable/writable

File: interop/reusability.md [hunk 3]
```diff
@@ -0,0 +1,1088 @@
...
+  configurable: boolean,    // default: true
+  writable: boolean,        // default: true
+  property: MethodProperty<Class>
+}
+
+interface AccessorDecoratorDescriptor<Class> {
+  type: string,             // 'accessor'
+  hint: string,             // 'getter', 'setter', 'both'
+  enumerable: boolean,      // default: true
+  wrtiable: boolean,        // default: true
```
Comment (jayphelps) [inline interop/reusability.md:973]: s/wrtiable/writable

File: interop/reusability.md [hunk 4]
```diff
@@ -0,0 +1,1088 @@
...
+  property: Property<Class>
+}
+```
+
+---
+
+```ts
+interface PropertyDecoratorDescriptor<Class> extends DecoratorDescriptor<Class> {
+  type: string,             // 'property'
+  hint: string,             // 'shorthand' or 'explicit' or 'static'
```
Comment (jayphelps) [inline interop/reusability.md:946]: What happens if it's a `PropertyDecoratorDescriptor` that is an accessor? It appears you do not receive any hints about which ones it contains like `AccessorDecoratorDescriptor`'s `"getter" | "setter" | "both"`, and `desc.type !==  "accessor"`. Seems inconsistent and confusing. If you instead return `AccessorDecoratorDescriptor` then they miss out on the hint of whether it's static.

``` js
const obj = {
  @foo
  get bar() {}
};
```

**Comment by rbuckton** [comment]
A few years ago, when I first discussed decorators with @wycats, I started with something very similar to the `DecoratorDescriptor` (nee. `DecoratorMetadata`), and was talked down from it. Is there a reason why we suddenly need this more complicated form of a property descriptor?

It seems like it might be simpler to continue to use Property Descriptors, and have some helper APIs on Reflect to get information about the kind of thing being decorated, via context.

``` js
// just using property descriptors
function dec(target, key, descriptor) {
  if (typeof descriptor === "object") {
    if (typeof descriptor.get === "function" || typeof descriptor.set === "function") {
      // accessor
    }
    else if (typeof descriptor.value === "function") {
      // method
    }
    else {
      // field
    }
  }
  else if (typeof target === "function") {
    // class constructor or function decorator
  }
}

// with an api
function dec(target, key, descriptor) {
  let kind = Reflect.getDescriptorKind(descriptor);
  if (kind === "accessor") {
    // accessor
  }
  else if (kind === "method") {
    // method
    // or possibly "data" since we already distinguish accessor properties vs data properties in the spec,
    // but then you'd have to special case a "data" property that has a function for its value.
  }
  else if (kind === "initializer") {
    // property
    // get initializer via `target[Symbol.classProperties][name]`?
    // or descriptor.initializer?
  }
  else if (typeof target === "function") {
    // class constructor/function
  }
}
```

After so much time being convinced that a new, specialized descriptor is a _bad idea_, this feels like a step backwards.

**Comment by rbuckton** [comment]
@wycats Is there an up to date strawman/proposal for private slots with any information as to the requirements for private slots around information disclosure? I thought the goal for private slots was to have private state that doesn't leak without some explicit public API from the class (such as an accessor around the private state). While a decorator generally could be thought of as explicit, I worry that providing any mechanism to read/write the private state via the descriptor passed to the decorator could be considered a risk in the eyes of the folks who want to avoid avenues of disclosing private state, since those decorators are external to the class body and have the possibility of being hijacked.

Without a clear direction on the goals for private slots, I'd be somewhat apprehensive about designing decorators around a feature that may be explicitly disallowed.

**Comment by silkentrance** [comment]
@wycats will this be made available publicly soon? I am considering migrating to babel 6.x...

**Comment by jayphelps** [comment]
@silkentrance this repo contains no actual code, it's just the draft specification unrelated to Babel.

You can track progress on Babel 6 supporting decorators here https://phabricator.babeljs.io/T2645

Someone released a compatibility plugin in that thread as well.

**Comment by trikadin** [comment]
@silkentrance you can use this plagin while the new version of this spec (and the new babel plugin) is not ready:
https://www.npmjs.com/package/babel-plugin-transform-decorators-legacy

**Comment by silkentrance** [comment]
@trikadin thanks for the pointer. will use this for now.

**Pull request merged**